File: emu/emu_plot/python/plot_fgrd.py
# Read Forward Gradient Tool output

import os
import numpy as np
import glob
import matplotlib.pyplot as plt
import global_emu_var as emu
import lib_python as emupy

# Create a custom colormap that maps masked values to gray
cmap = plt.cm.jet
cmap.set_bad(color='gray')  # Set the color for masked elements to gray

def plot_fgrd(frun): 
    # Define the byte order ('>' for big-endian)
    byte_order = '>'

    # Set EMU output file directory
    ff = os.path.join(frun, 'output')


    # ---------------
    # Search available output

    print()
    print('Detected ')

    # Search for files matching the pattern
    fdum = 'state_2d_set1_day.*.data'
    aa_2d_day = glob.glob(os.path.join(ff, fdum))
    naa_2d_day = len(aa_2d_day)
    print(f"{naa_2d_day:6d} files of {fdum}")

    fdum = 'state_2d_set1_mon.*.data'
    aa_2d_mon = glob.glob(os.path.join(ff, fdum))
    naa_2d_mon = len(aa_2d_mon)
    print(f"{naa_2d_mon:6d} files of {fdum}")

    fdum = 'state_3d_set1_mon.*.data'
    aa_3d_mon = glob.glob(os.path.join(ff, fdum))
    naa_3d_mon = len(aa_3d_mon)
    print(f"{naa_3d_mon:6d} files of {fdum}")

    # ---------------
    # Define gradient variable to read and plot 
    f_var = ['SSH', 'OBP', 'THETA', 'SALT', 'U', 'V']
    nvar = len(f_var)

    # Prompt user to choose a variable
    print("\nChoose variable to plot ...")
    for i in range(nvar):
        pdum = f"{i+1}) {f_var[i]}"
        print(pdum)

    # Prompt user to select monthly or daily mean
    print("\nSelect monthly or daily mean ... (m/d)?")
    print("(NOTE: daily mean available for SSH and OBP only.)")
    fmd = input("Enter 'm' for monthly or 'd' for daily: ").strip().lower()

    # ---------------
    # Read and plot gradient of chosen variable 
    pvar = 0
    idum = 0

    if fmd == 'd':
        print("==> Reading and plotting daily means ...\n")
        while pvar < 1 or pvar > 2:
            pvar = int(input("Enter variable # to plot ... (1-2)? "))
            ivar = pvar - 1

        print(f"\nPlotting ... {f_var[ivar]}")

        # loop among daily mean 2d files 
        pfile = 1
        while pfile >= 1 and pfile <= naa_2d_day:
            print()
            pfile = int(input(f"Enter file # to read ... (1-{naa_2d_day})?"))
            if pfile < 1 or pfile > naa_2d_day:
                break
            ifile = pfile-1

            emu.fgrd2d = emupy.rd_state2d_r4(aa_2d_day[ifile], ivar)
            fname = os.path.basename(aa_2d_day[ifile])
            pinfo = f"{f_var[ivar]} {pfile} {fname}"
            emupy.plt_state2d(emu.fgrd2d, pinfo)

        print('*********************************************')
        print('Returning variable ')
        print('   fgrd2d: last plotted gradient (2d)')
        print()

    else:
        print("==> Reading and plotting monthly means ...\n")
        while pvar < 1 or pvar > nvar:
            pvar = int(input(f"Enter variable # to plot ... (1-{nvar})? "))
            ivar = pvar - 1

        print(f"\nPlotting ... {f_var[ivar]}")

        if ivar <= 1:

        # loop among monthly mean 2d files 
            pfile = 1
            while pfile >= 1 and pfile <= naa_2d_mon:
                print()
                pfile = int(input(f"Enter file # to read ... (1-{naa_2d_mon})?"))
                if pfile < 1 or pfile > naa_2d_mon:
                    break
                ifile = pfile-1
                
                emu.fgrd2d = emupy.rd_state2d_r4(aa_2d_mon[ifile], ivar)
                fname = os.path.basename(aa_2d_mon[ifile])
                pinfo = f"{f_var[ivar]} {pfile} {fname}"
                emupy.plt_state2d(emu.fgrd2d, pinfo)

            print('*********************************************')
            print('Returning variable ')
            print('   fgrd2d: last plotted gradient (2d)')
            print()

        else:

        # loop among monthly mean 3d files 
            pfile = 1
            while pfile >= 1 and pfile <= naa_3d_mon:
                print()
                pfile = int(input(f"Enter file # to read ... (1-{naa_3d_mon})?"))
                if pfile < 1 or pfile > naa_3d_mon:
                    break
                ifile = pfile-1

                emu.fgrd3d = emupy.rd_state3d(aa_3d_mon[ifile], ivar-2) # ivar-2 because no 3d SSH/OBP
                fname = os.path.basename(aa_3d_mon[ifile])
                pinfo = f"{f_var[ivar]} {pfile} {fname}"
                emupy.plt_state3d(emu.fgrd3d, pinfo, ivar-2) # ivar-2 because no 3d SSH/OBP

            print('*********************************************')
            print('Returning variable ')
            print('   fgrd3d: last plotted gradient (3d)')
            print()

File: emu/emu_plot/matlab/plot_fgrd.m
function plot_fgrd(frun)
    % PLOT_FGRD  Read and plot EMU forward gradient tool output
    %
    %   Input:
    %     frun - path to EMU run directory (e.g., 'emu_fgrd_7_15_743_124_-1.00E-0')
    global emu

    % Set EMU output directory
    ff = fullfile(frun, 'output');

    % Search available output files
    aa_2d_day = dir(fullfile(ff, 'state_2d_set1_day.*.data'));
    aa_2d_mon = dir(fullfile(ff, 'state_2d_set1_mon.*.data'));
    aa_3d_mon = dir(fullfile(ff, 'state_3d_set1_mon.*.data'));

    naa_2d_day = numel(aa_2d_day);
    naa_2d_mon = numel(aa_2d_mon);
    naa_3d_mon = numel(aa_3d_mon);

    fprintf('%6d files of state_2d_set1_day\n', naa_2d_day);
    fprintf('%6d files of state_2d_set1_mon\n', naa_2d_mon);
    fprintf('%6d files of state_3d_set1_mon\n', naa_3d_mon);

    % Define variable list
    f_var = {'SSH', 'OBP', 'THETA', 'SALT', 'U', 'V'};
    nvar = numel(f_var);

    fprintf('\nVariables to plot ...\n');
    for i = 1:nvar
        fprintf('%d) %s\n', i, f_var{i});
    end

    fmd = lower(input('\nSelect monthly or daily mean (m/d)? ', 's'));

    if strcmp(fmd, 'd')
        fprintf('==> Reading and plotting daily means ...\n');
        pvar = 0;
        while pvar < 1 || pvar > 2
            pvar = input('Enter variable # to plot ... (1-2)? ');
        end
        ivar = pvar;

        while true
            pfile = input(sprintf('Enter file # to read ... (1-%d or -1 to exit)? ', naa_2d_day));
            if pfile < 1 || pfile > naa_2d_day
                break;
            end
            fname = fullfile(ff, aa_2d_day(pfile).name);
            fgrd2d = rd_state2d_r4(fname, ivar);
            title_str = sprintf('%s %d %s', f_var{ivar}, pfile, aa_2d_day(pfile).name);
            plt_state2d(fgrd2d, title_str);
        end
	fprintf('\n*********************************************')
        fprintf('\nReturning variable: fgrd2d (last plotted gradient 2D)\n');
	emu.fgrd2d = fgrd2d;

    else
        fprintf('==> Reading and plotting monthly means ...\n');
        pvar = 0;
        while pvar < 1 || pvar > nvar
            pvar = input(sprintf('Enter variable # to plot ... (1-%d)? ', nvar));
        end
        ivar = pvar;

        if ivar <= 2  % 2D case
            while true
                pfile = input(sprintf('Enter file # to read ... (1-%d or -1 to exit)? ', naa_2d_mon));
                if pfile < 1 || pfile > naa_2d_mon
                    break;
                end
                fname = fullfile(ff, aa_2d_mon(pfile).name);
                fgrd2d = rd_state2d_r4(fname, ivar);
                title_str = sprintf('%s %d %s', f_var{ivar}, pfile, aa_2d_mon(pfile).name);
                plt_state2d(fgrd2d, title_str);
            end
	    fprintf('\n*********************************************')
            fprintf('\nReturning variable: fgrd2d (last plotted gradient 2D)\n');
	    emu.fgrd2d = fgrd2d;

        else  % 3D case
            while true
                pfile = input(sprintf('Enter file # to read ... (1-%d or -1 to exit)? ', naa_3d_mon));
                if pfile < 1 || pfile > naa_3d_mon
                    break;
                end
                fname = fullfile(ff, aa_3d_mon(pfile).name);
                fgrd3d = rd_state3d(fname, ivar - 2);  % no 3D SSH/OBP
                title_str = sprintf('%s %d %s', f_var{ivar}, pfile, aa_3d_mon(pfile).name);
                plt_state3d(fgrd3d, title_str, ivar - 2);
            end
	    fprintf('\n*********************************************')
            fprintf('\nReturning variable: fgrd3d (last plotted gradient 3D)\n');
	    emu.fgrd3d = fgrd3d;
        end
    end
end
